firstvoices/backend/views/base_views.py

```diff
@@ -1,25 +1,113 @@
-class FVPermissionViewSetMixin(AutoPermissionViewSetMixin):
+
+class FVPermissionViewSetMixin:
     """
+    Forked from ``rules.contrib.rest_framework.AutoPermissionViewSetMixin`` to provide extension points.
+
     Enforces object-level permissions in ``rest_framework.viewsets.ViewSet``,
     deriving the permission type from the particular action to be performed. List results are
     filtered to only include permitted items.
 
     As with ``rules.contrib.views.AutoPermissionRequiredMixin``, this only works when
     model permissions are registered using ``rules.contrib.models.RulesModelMixin``.
+
+    Override ``get_object_for_create_permission`` to check 'create' permissions
+    on an object if desired (defaults to a None object).
     """
 
+    # Maps API actions to model permission types. None as value skips permission
+    # checks for the particular action.
+    # This map needs to be extended when custom actions are implemented
+    # using the @action decorator.
+    # Extend or replace it in subclasses like so:
+    # permission_type_map = {
+    #     **AutoPermissionViewSetMixin.permission_type_map,
+    #     "close": "change",
+    #     "reopen": "change",
+    # }
+    permission_type_map = {
+        "create": "add",
+        "destroy": "delete",
+        "list": None,
+        "partial_update": "change",
+        "retrieve": "view",
+        "update": "change",
+        "get_preview": None,
+        "create_preview": "add",
+    }
+
+    def initial(self, *args, **kwargs):
+        """Ensures user has permission to perform the requested action."""
+        super().initial(*args, **kwargs)
+
+        if not self.request.user:
+            # No user, don't check permission
+            return
+
+        # Get the handler for the HTTP method in use
+        try:
+            if self.request.method.lower() not in self.http_method_names:
+                raise AttributeError
+            handler = getattr(self, self.request.method.lower())
+        except AttributeError:
+            # method not supported, will be denied anyway
+            return
+
+        try:
+            perm_type = self.permission_type_map[self.action]
+        except KeyError:
+            raise ImproperlyConfigured(
+                "FVPermissionViewSetMixin tried to authorize a request with the "
+                "{!r} action, but permission_type_map only contains: {!r}".format(
+                    self.action, self.permission_type_map
+                )
+            )
+        if perm_type is None:
+            # Skip permission checking for this action
+            return
+
+        # Determine whether we've to check object permissions (for detail actions)
+        obj = None
+        extra_actions = self.get_extra_actions()
+        # We have to access the unbound function via __func__
+        if handler.__func__ in extra_actions:
+            if handler.detail:
+                obj = self.get_object()
+            elif "create" in self.action:
```

Comment:
I made this change so that any additional action contain the word "create" would be able to use the `get_object_for_create_permission` function. In my case it would be fore "create_preview" but it could apply to other additional actions in the future.

Reply:
Oh, gotcha. Good point about custom permissions. This is fine for now, but I am pondering how we might handle other permission names in the future.